web3/types.py

```diff
@@ -37,6 +37,11 @@
 TValue = TypeVar("TValue")
 
 
+HexBytes32 = NewType("HexBytes32", HexBytes)
+HexStr32 = NewType("HexStr32", HexStr)
+
+_Hash32 = Union[Hash32, HexBytes, HexStr]
```

Comment:
@pipermerriam I just realized that I've most likely been using the `eth_typing.Hash32` type incorrectly, which is defined as a `bytes` type of len 32. I assumed it was any string or bytes representing a 32 byte hash. I think it would be nice to have a type that consolidates `Union[Hash32, HexBytes, HexStr]` which are all valid throughout `web3`.

One thought is maybe introducing `Bytes32` to replace `Hash32`, and then `Hash32` can be used to represent `Union[Bytes32, HexBytes32, HexStr32]` - But it's unclear how rigid the `Hash32` definition is used in `py-evm`.

Reply:
`Hash32` is used very broadly across `py-evm` and `trinity` and this would be a breaking change to `eth-typing` which sounds like a logistical nightmare.  I'm not categorically opposed but the upgrade path seems riddled with insanity.